plugins/inputs/elasticsearch_query/aggregation_parser.go

```diff
@@ -0,0 +1,154 @@
+}
+
+func (e *ElasticsearchQuery) parseSimpleResult(measurement string, searchResult *elastic.SearchResult, acc telegraf.Accumulator) error {
+
+	fields := make(map[string]interface{})
+	tags := make(map[string]string)
+
+	fields["doc_count"] = searchResult.Hits.TotalHits
+
+	acc.AddFields(measurement, fields, tags)
+	return nil
+}
+
+func (e *ElasticsearchQuery) parseAggregationResult(aggregationQueryList []aggregationQueryData, searchResult *elastic.SearchResult, acc telegraf.Accumulator) error {
+	var measurements = map[string]map[string]string{}
+
+	// organize the aggregation query data by measurement
+	for _, aggregationQuery := range aggregationQueryList {
+
+		if measurements[aggregationQuery.measurement] == nil {
+			measurements[aggregationQuery.measurement] = map[string]string{
+				aggregationQuery.name: aggregationQuery.function,
+			}
+		} else {
+			t := measurements[aggregationQuery.measurement]
+			t[aggregationQuery.name] = aggregationQuery.function
+			measurements[aggregationQuery.measurement] = t
+		}
+	}
+
+	// recurse over aggregation results per measurement
+	for measurement, aggNameFunction := range measurements {
+		var m resultMetric
+
+		m.fields = make(map[string]interface{})
+		m.tags = make(map[string]string)
+		m.name = measurement
+
+		m, err := e.recurseResponse(aggNameFunction, searchResult.Aggregations, m, acc)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (e *ElasticsearchQuery) recurseResponse(aggKeys map[string]string, bucketResult elastic.Aggregations, m resultMetric, acc telegraf.Accumulator) (resultMetric, error) {
+	var err error
+
+	aggName, found := getAggName(bucketResult)
+	if !found {
+		// we've reached a single bucket without aggregation, nothing here
+		return m, nil
+	}
+
+	aggFunction, found := aggKeys[aggName]
+	if !found {
+		return m, fmt.Errorf("child aggregation function '%s' not found %v", aggName, aggKeys)
+	}
+
+	resp := e.getResponseAggregation(aggFunction, aggName, bucketResult)
+	if resp == nil {
+		return m, fmt.Errorf("child aggregation '%s' not found", aggName)
+	}
+
+	switch resp := resp.(type) {
+	case *elastic.AggregationBucketKeyItems:
+		// we've found a terms aggregation, iterate over the buckets and try to retrieve the inner aggregation values
+		for _, bucket := range resp.Buckets {
+			m.fields["doc_count"] = bucket.DocCount
+			if s, ok := bucket.Key.(string); ok {
+				m.tags[aggName] = s
+			} else {
+				return m, fmt.Errorf("bucket key is not a string")
+			}
+
+			// we need to recurse down through the buckets, as it may contain another terms aggregation
+			m, err = e.recurseResponse(aggKeys, bucket.Aggregations, m, acc)
+			if err != nil {
+				return m, err
+			}
+
+			// if there are fields present after finishing the bucket, it is a complete metric
+			// store it and clean the fields to start a new metric
+			if len(m.fields) > 0 {
+				acc.AddFields(m.name, m.fields, m.tags)
+				m.fields = make(map[string]interface{})
+			}
```

Comment:
I think you need a flag to denote that you are at the highest level again before adding the fields. Imagine you go down the tree, say 4 times, then you will add at least 3 times the same metric with a growing number of fields. Remember fields are non-zero in every level except maybe the lowest.
So either add a depth counter and only add the metric if you are at depth zero or split the recursion part from the top-level function...

Reply:
The metric needs to be added after it reaches the highest level. 

Until it gets there, the tags are being added to the metric, and the fields too, except that the `doc_count` field is overwritten each time it goes up a level to another terms aggregation. Only when it reaches the `metric` aggregation (at the highest level of the recursion) is that it adds other field names like `field_avg` or `field_max`.

So I'm adding the metric when it goes down the first time. By resetting the `fields` and removing that bucket `tag` from the metric struct it means that it will not duplicate that metric when continuing to next bucket (new tag, new field) or going down the recursion. 

As the `metric` struct is returned each time from the higher recursion function down before checking if it needs to be added, normally it will not add again when going down the next time. 

By looking now I need to confirm if it is working correctly when more than one metric field is present.
It seems I will need to add the metric and reset fields at the end of the `AggregationValueMetric` switch case; and the current `addFields` would only be called when no metric field is present.

Reply:
Why don't you collect _all_ fields and _all_ tags during recursion and the call `acc.AddFields(m.name, m.fields, m.tags)` in the measurement loop of `parseAggregationResult()`? I think this will be much easier to understand... I'm still unsure if your description above matches what is done in the code after reading this function multiple times now. :-)

Reply:
I could do that, so instead of calling `acc.AddFields` directly I could add the resulting metric to a map (or something else) and iterate over that in `parseAggregationResult`, but I think that would be more complex. 

Then I would need to see how to work with the tags slice inside the `metric` struct as I need to add/remove values from it during the recursive calls. As slices work by reference it would change the values in the resulting map as well (I figured that out because I tried that).

I want to add tests to this using JSON responses from these aggregations in Elasticsearch. Maybe by looking at these will help the understanding of this particular part of the code (and maybe to improve with a better solution).

Reply:
To be honest it cannot be as hard to read as this. I _really_ tried to wrap my head around it and I'm still not sure I get all the possible paths, so splitting this into two parts might be a good idea. :-)
I still don't get the idea I think. As far as I get it you get a response that has a tree like structure. Is this correct? That's why you recurse into it and build aggregations on the way. However, you want to collect the whole subtree to form a metric, don't you?
Your code suggest that you call this recursions for _each_ **measurement**, so the first question is, do you expect multiple metrics for each measurement? If yes, do you have a clear signal of when a metric ends and another starts? Currently it looks like you want a metric for each `bucket` and each `metric`, but only at the highest level, all other things further down the tree should be collected/aggregated. Is this correct?

If all of the above is true, moving `acc.AddFields()` to `parseAggregationResult()` is a bad idea. In this case I would suggest to pass a `depth` counter down the recursive function and only call  `acc.AddFields()` if that counter is zero.

Reply:
> To be honest it cannot be as hard to read as this. I really tried to wrap my head around it and I'm still not sure I get all the possible paths, so splitting this into two parts might be a good idea. :-)

It was painful for me to go back to this code too :)

> I still don't get the idea I think. As far as I get it you get a response that has a tree like structure. Is this correct? 

Yes, exactly, and the objective is somehow to flat that tree.

> That's why you recurse into it and build aggregations on the way. However, you want to collect the whole subtree to form a metric, don't you?

Yes.

> Your code suggest that you call this recursions for _each_ **measurement**, so the first question is, do you expect multiple metrics for each measurement?

Yes, that's the case

> ... do you have a clear signal of when a metric ends and another starts? Currently it looks like you want a metric for each `bucket` and each `metric`, but only at the highest level, all other things further down the tree should be collected/aggregated. Is this correct?

Yes. The metric stops at the highest level as you mention. The highest level is a metric aggregation, when it exists. But a metric aggregation is optional, and the metric may end at the highest level for each bucket.

The metric aggregation (when it exists) is always at the edge of the (sub)tree. 

> If all of the above is true, moving `acc.AddFields()` to `parseAggregationResult()` is a bad idea. In this case I would suggest to pass a `depth` counter down the recursive function and only call `acc.AddFields()` if that counter is zero.

I will investigate this during the weekend if I can, but my feeling is that if the counter is zero it is too late to cal the `acc.AddFields()` as the tags are changing if there are more buckets. The part that I'm really in doubt is about calling it inside the metric aggregation though, I need to test the code using at least 3 terms aggregation and 2 metrics aggregation (the current test data only allowed me 1 metrics aggregation).

Reply:
Thanks for your effort but I think it's worth it... :-)